src/brod_utils.erl

```diff
@@ -237,8 +277,195 @@ init_sasl_opt(Config) ->
+    fun(Pid) -> fetch_committed_offsets(Pid, GroupId) end).
+
+%% @doc Same as `fetch_committed_offsets/3', only work on a socket
+%% connected to the group coordinator broker.
+%% @end
+-spec fetch_committed_offsets(pid(), group_id()) ->
+        {ok, [kpro:struct()]} | {error, any()}.
+fetch_committed_offsets(SockPid, GroupId) ->
+  Body = [ {group_id, GroupId}
+         , {topics, ?undef} %% Fetch ALL when topics array is null
+         ],
+  %% {topics, ?undef} works only with version 2
+  Vsn = 2,
+  Req = kpro:req(offset_fetch_request, Vsn, Body),
+  try
+    #kpro_rsp{ tag = offset_fetch_response
+             , vsn = Vsn
+             , msg = Msg
+             } = request_sync(SockPid, Req),
+    TopicsArray = kf(responses, Msg),
+    {ok, TopicsArray}
+  catch
+    throw : Reason ->
+      {error, Reason}
+  end.
+
 %%%_* Internal Functions =======================================================
 
+%% @private Make a function to build fetch requests.
+%% The function takes offset and max_bytes as input as these two parameters
+%% are varient when continuously polling a specific topic-partition.
+%% @end
+-spec make_req_fun(pid(), topic(), partition(),
+                   kpro:wait(), kpro:count()) -> req_fun().
+make_req_fun(_SockPid, Topic, Partition, WaitTime, MinBytes) ->
+  Vsn = 0, %% TODO pick version (make use of SockPid)
+  fun(Offset, MaxBytes) ->
+      kpro:fetch_request(Vsn, Topic, Partition, Offset,
+                         WaitTime, MinBytes, MaxBytes)
+  end.
+
+%% @doc Fetch a message-set. If the given MaxBytes is not enough to fetch a
+%% single message, expand it to fetch exactly one message
+%% @end
+-spec fetch(pid(), req_fun(), offset(), kpro:count()) ->
+               {ok, [brod:message()]} | {error, any()}.
+fetch(SockPid, ReqFun, Offset, MaxBytes) when is_pid(SockPid) ->
+  Request = ReqFun(Offset, MaxBytes),
+  #kpro_rsp{ tag = fetch_response
+           , msg = Msg
+           } = request_sync(SockPid, Request, infinity),
+  [Response] = kf(responses, Msg),
+  [PartitionResponse] = kf(partition_responses, Response),
+  Header = kf(partition_header, PartitionResponse),
+  Messages0 = kf(record_set, PartitionResponse),
+  ErrorCode = kf(error_code, Header),
+  case ?IS_ERROR(ErrorCode) of
+    true ->
+      {error, kpro_error_code:desc(ErrorCode)};
+    false ->
+      case decode_messages(Offset, Messages0) of
+        ?incomplete_message(Size) ->
+          fetch(SockPid, ReqFun, Offset, Size);
+        Messages ->
+          {ok, Messages}
+      end
+  end.
+
+%% @doc List all groups in the given cluster.
+%% NOTE: Exception if failed against any of the coordinator brokers.
+%% @end
+-spec list_all_groups([endpoint()], sock_opts()) ->
+        [{endpoint(), [brod:cg()] | {error, any()}}].
+list_all_groups(Endpoints, Options) ->
+  {ok, Metadata} = get_metadata(Endpoints, [], Options),
+  Brokers0 = kf(brokers, Metadata),
+  Brokers = [{binary_to_list(kf(host, B)), kf(port, B)} || B <- Brokers0],
```

Comment:
I tried. However Hostname being a string() is assertively used in many places.
It is also used as ETS keys.
Doesn't gain much to support `binary() | string()` spec everywhere.

I'd prefer to support a `raw_hostname() :: binary() | string()` for public APIs
and ensure `hostname() :: string()` everywhere internally.